src/bp/core/services/auth/basic.ts

```diff
@@ -0,0 +1,232 @@
+    const password = nanoid(15)
+    const { hash, salt } = saltHashPassword(password)
+
+    await this.authService.updateUser(email, strategy, { password: hash, salt })
+    await this.authService.updateAttributes(email, strategy, { password_expired: true })
+
+    debug('password reset %o', { email })
+
+    return password
+  }
+
+  private async _login(
+    email: string,
+    password: string,
+    strategy: string,
+    newPassword?: string,
+    ipAddress: string = ''
+  ): Promise<void> {
+    await this._checkUserAuth(email, strategy, password, newPassword, ipAddress)
+    const strategyOptions = _.get(await this.authService.getStrategy(strategy), 'options') as AuthStrategyBasic
+
+    if (newPassword) {
+      this._validatePassword(newPassword, strategyOptions)
+      const hash = saltHashPassword(newPassword)
+
+      await this.authService.updateUser(email, strategy, {
+        password: hash.hash,
+        salt: hash.salt,
+        ...this._addPasswordExpiry(strategyOptions)
+      })
+
+      await this.authService.updateAttributes(email, strategy, { password_expired: false })
+    }
+
+    debug('login', { email, ipAddress })
+
+    await this.authService.updateAttributes(email, strategy, {
+      last_ip: ipAddress,
+      unsuccessful_logins: 0,
+      last_login_attempt: undefined,
+      locked_out: false,
+      last_logon: new Date()
+    })
+  }
+
+  private async _register(email: string, strategy: string, password: string, ipAddress: string = ''): Promise<string> {
+    const pw = saltHashPassword(password)
+
+    if (await this.authService.findUser(email, strategy)) {
+      throw new ConflictError('The user already exists for that strategy.')
+    }
+
+    const newUser = {
+      email,
+      password: pw.hash,
+      salt: pw.salt,
+      strategy,
+      attributes: {
+        last_ip: ipAddress,
+        last_logon: new Date()
+      }
+    }
+
+    await this.authService.createUser(newUser, strategy)
+
+    debug('self register', { email, ipAddress })
+    return this.authService.generateSecureToken(email, strategy)
+  }
+
+  private async _checkUserAuth(
+    email: string,
+    strategy: string,
+    password: string,
+    newPassword?: string,
+    ipAddress: string = ''
+  ) {
+    if (email === SERVER_USER) {
+      debug('user tried to login with server user %o', { email, ipAddress })
+      throw new InvalidCredentialsError()
+    }
+
+    const user = await this.authService.findUser(email, strategy)
+    if (!user) {
+      debug('login failed; user does not exist %o', { email, ipAddress })
+      throw new InvalidCredentialsError()
+    }
+    const strategyOptions = _.get(await this.authService.getStrategy(strategy), 'options') as AuthStrategyBasic
+    if (!validateHash(password || '', user.password!, user.salt!)) {
+      debug('login failed; wrong password %o', { email, ipAddress })
+      // this.stats.track('auth', 'login', 'fail')
```

Comment:
should we keep this ?

Reply:
I'm not sure, so I commented it. If we want to add the stats, I wasn't sure where to add them, and wanted to add them phase 2